packages/react-testing/src/root.tsx

```diff
@@ -0,0 +1,249 @@
+import * as React from 'react';
+import {render, unmountComponentAtNode} from 'react-dom';
+import {act} from 'react-dom/test-utils';
+import {
+  Arguments,
+  Props as PropsForComponent,
+  MaybeFunctionReturnType as ReturnType,
+} from '@shopify/useful-types';
+
+import {TestWrapper} from './TestWrapper';
+import {Element, Predicate} from './element';
+import {Tag, Fiber, ReactInstance, FunctionKeys} from './types';
+import {withIgnoredReactLogs} from './errors';
+
+// eslint-disable-next-line typescript/no-var-requires
+const {findCurrentFiberUsingSlowPath} = require('react-reconciler/reflection');
+
+export const connected = new Set<Root<unknown>>();
+
+export class Root<Props> {
+  get props() {
+    return this.withRoot(root => root.props);
+  }
+
+  get isDOM() {
+    return this.withRoot(root => root.isDOM);
+  }
+
+  get instance() {
+    return this.withRoot(root => root.instance);
+  }
+
+  get children() {
+    return this.withRoot(root => root.children);
+  }
+
+  get descendants() {
+    return this.withRoot(root => root.descendants);
+  }
+
+  get domNodes() {
+    return this.withRoot(root => root.domNodes);
+  }
+
+  get domNode() {
+    return this.withRoot(root => root.domNode);
+  }
+
+  private wrapper: TestWrapper<Props> | null = null;
+  private element = document.createElement('div');
+  private root: Element<Props> | null = null;
+
+  private get mounted() {
+    return this.wrapper != null;
+  }
+
+  constructor(private tree: React.ReactElement<Props>) {
+    this.mount();
+  }
+
+  perform<T>(action: () => T, {update = true} = {}): T {
+    let result!: T;
+
+    withIgnoredReactLogs(() =>
+      act(() => {
+        result = action();
+      }),
+    );
+
+    if (update) {
+      this.update();
+    }
+
+    return result;
+  }
+
+  html() {
+    return this.withRoot(root => root.html());
+  }
+
+  text() {
+    return this.withRoot(root => root.text());
+  }
+
+  is<Type extends React.ComponentType<any> | string>(
+    type: Type,
+  ): this is Root<PropsForComponent<Type>> {
+    return this.withRoot(root => root.is(type));
+  }
+
+  prop<K extends keyof Props>(key: K) {
+    return this.withRoot(root => root.prop(key));
+  }
+
+  find<Type extends React.ComponentType<any> | string>(type: Type) {
+    return this.withRoot(root => root.find(type));
+  }
+
+  findAll<Type extends React.ComponentType<any> | string>(type: Type) {
```

Comment:
Can we deviate from the enzyme APIs and closer to common functional APIs for this stuff?  Personally I think it would be nice to have like `find`, `filter` instead of `find`, `findAll`

Reply:
This is a tough one. Personally, I find `filter` to be confusing, mostly because some methods return arrays, and arrays have their own filter. `filter` also sounds like the object is a collection (which is not the case). Technically `find` should have all the same problems but in my brain it doesn't get associated with arrays in the same way. Maybe I am crazy?

Reply:
I'm having trouble responding to this because I like that some methods return arrays that have their own filter :P 

It's not an array but semantically `filter` and `find` make just as much sense as in an array IMO.

Not something I feel *that* strongly about, I just always found enzyme's choice here a bit weird.